atlasdb-timelock-server/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStoreTest.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    @Test
+    public void throwsIfBoundUnexpectedlyChangedUnderUs() {
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(1);
+        additionalStore.storeUpperLimit(TIMESTAMP_1);
+        assertThatThrownBy(() -> store.storeUpperLimit(TIMESTAMP_2))
+                .isInstanceOf(MultipleRunningTimestampServiceError.class);
+    }
+
+    @Test
+    public void canReadStateFromDistributedLogs() {
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(1);
+        additionalStore.storeUpperLimit(TIMESTAMP_1);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_1);
+        store.storeUpperLimit(TIMESTAMP_2);
+        assertThat(store.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_2);
+    }
+
+    @Test
+    public void canReadStateThroughLeaderChanges() {
+        PaxosTimestampBoundStore additionalStore1 = createPaxosTimestampBoundStore(1);
+        PaxosTimestampBoundStore additionalStore2 = createPaxosTimestampBoundStore(0);
+
+        store.storeUpperLimit(TIMESTAMP_1);
+        assertThat(additionalStore1.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_1);
+        additionalStore1.storeUpperLimit(TIMESTAMP_2);
+        assertThat(additionalStore2.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_2);
+        additionalStore2.storeUpperLimit(TIMESTAMP_3);
+        assertThat(additionalStore2.getUpperLimit()).isGreaterThanOrEqualTo(TIMESTAMP_3);
+    }
+
+    @Test
+    public void canGetAgreedInitialState() {
+        PaxosTimestampBoundStore.SequenceAndBound sequenceAndBound = store.getAgreedState(0);
+        assertThat(sequenceAndBound.getSeqId()).isEqualTo(0);
+        assertThat(sequenceAndBound.getBound()).isEqualTo(0);
+    }
+
+    @Test
+    public void canGetAgreedState() {
+        store.storeUpperLimit(TIMESTAMP_1);
+        PaxosTimestampBoundStore.SequenceAndBound sequenceAndBound = store.getAgreedState(1);
+        assertThat(sequenceAndBound.getSeqId()).isEqualTo(1);
+        assertThat(sequenceAndBound.getBound()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void canSafelyGetAgreedStateFromPrehistory() {
+        assertThat(store.getAgreedState(Long.MIN_VALUE).getBound()).isEqualTo(0);
+    }
+
+    @Test
+    public void canGetAgreedStateAfterNodeDown() {
+        failureToggles.get(1).set(true);
+        store.storeUpperLimit(TIMESTAMP_1);
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(1);
+        failureToggles.get(1).set(false);
+
+        assertThat(additionalStore.getAgreedState(2).getBound()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void cannotGetAgreedStateFromTheFuture() {
+        assertThatThrownBy(() -> store.getAgreedState(Long.MAX_VALUE))
+                .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void canSafelyForceAgreedStateFromPrehistory() {
```

Comment:
You *can*, though it's semantically awkward for Paxos. `Force(seq, num)` basically behaves as follows:

* If consensus for `seq` is already known by any learner, return it
* Otherwise, have the proposer propose `num`.
* The proposer will return a value that a quorum agreed on (maybe `num`, maybe not).
* We return this value.

If `num` is null (this is a signal that we want to find out a value, as opposed to setting it) and the cluster hasn't agreed on a value, we fail. Using null for `num` is typically done only if learner state is lost.

Multi-Paxos allows there to be gaps in state as far as an individual node is concerned (though that's from my understanding of the algorithm, not about our specific impl), so there's nothing _wrong_ with this behaviour in theory.

I've beefed up the tests for force(), in any case.